Add ingest script and dockerfile

nchevobbe opened the pull request.

nchevobbe committed:
Ignore node_modules changes

nchevobbe committed:
Add ingest script

nchevobbe committed:
Add DockerFile

juliandescottes commented on package.json:
```diff
@@ -0,0 +1,14 @@
+{
+  "name": "remote-settings-mdn-browser-data",
```
Should this match the name of the repo? `remote-settings-mdn-browser-compat-data`

nchevobbe replied:
sure

juliandescottes commented on package.json:
```diff
@@ -0,0 +1,14 @@
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "ingest": "node update_remote_settings_records.mjs"
+  },
+  "author": "",
```
Not sure how "complete" we should make this package json now that we have it on github. I guess we are not going to publish it, so not sure if a valid author and license are needed.

nchevobbe replied:
not sure either, I'll check if there's some convention about package.json file for projects which won't be published

nchevobbe replied:
I removed most of the properties, we can reevaluate if we ever need those in the future

nchevobbe commented:
@leplatrem can you have a look at the script and the Dockerfile?

leplatrem commented on update_remote_settings_records.mjs:
```diff
@@ -0,0 +1,352 @@
+// This script consumes env variables:
+// - AUTHORIZATION (mandatory): Raw authorization header (e.g. `AUTHORIZATION='Bearer XXXXXXXXXXXXX'`)
+// - SERVER (mandatory): Writer server URL (eg. https://settings-writer.stage.mozaws.net/v1)
+// - ENVIRONMENT (optional): dev, stage, prod. When set to `dev`, the data will be automatically published.
```
I would say *the script will approve its own changes*

leplatrem commented on update_remote_settings_records.mjs:
```diff
@@ -0,0 +1,352 @@
+ * Ask for review on the collection.
+ */
+async function requestReview() {
+  if (isDryRun) {
```
Maybe add a cnosole log here too

nchevobbe replied:
sure

leplatrem commented on update_remote_settings_records.mjs:
```diff
@@ -0,0 +1,352 @@
+ * ⚠️ This only works on the `dev` server.
+ */
+async function approveChanges() {
+  if (isDryRun) {
```
Here as well, log something on dry run

nchevobbe replied:
👍

leplatrem commented on update_remote_settings_records.mjs:
```diff
@@ -0,0 +1,352 @@
+// * css.properties: https://github.com/mdn/browser-compat-data/tree/main/css
+
+// The MDN compatibility data is available as a node package ("@mdn/browser-compat-data").
+// This node script fetches `browsers.json` and `css-properties.json` and updates records
```
`css-properties.json` is not yet handled :)

leplatrem commented on update_remote_settings_records.mjs:
```diff
@@ -0,0 +1,352 @@
+      "Content-Type": "application/json",
+      Authorization: process.env.AUTHORIZATION,
+    },
+  });
```
Instead of repeating these lines, you may want to do something like:

```js
async function fetchJSON(url, options = {}) {
  if (isDryRun && options.method != "GET") {
    console.log( ... );
    return;
  }
  const response = await fetch(url, {
    headers: {
      "Content-Type": "application/json",
      Authorization: process.env.AUTHORIZATION,
    },
    ...options
  });
  if (response.status >= 300) {
    ...
  }
  return await response.json();
}
```

nchevobbe replied:
I feel like this would complexify the code a bit for little returned value?
What we could do to avoid repetition is having a `headers` variable that we would instantiate once so we could do 

```js
  const response = await fetch(`${rsBrowsersRecordsEndpoint}/${record.id}`, {
    method: "PUT",
    body: JSON.stringify({ data: browserMdn }),
    headers,
  });
```

This looks not too different from 

```js
  const response = await fetchJSON(`${rsBrowsersRecordsEndpoint}/${record.id}`, {
    method: "PUT",
    body: JSON.stringify({ data: browserMdn }),
  });
```

while being more explicit and flexible for each functions (checking exact status code, retrieving returned json data only when needed, having proper warning message on api call failure)

leplatrem replied:
yes, great!

leplatrem reviewed (approved):
I had a few comments, but this is shippable :)

nchevobbe committed:
address review comments

prefix logs with DRY_RUN when env variable set.
putting logs in function making fetch calls.
creating shared headers variable.

nchevobbe committed:
reduce package.json to its minimum
